### DISCUSSION TITLE
Avram::Migrator::Runner.new.run_pending_migrations not work anymore?

### DISCUSSION THREAD

**Discussion by zw963** [luckyframework/lucky#1867] [Q&A] [answered]
Hi, i am back!

What i want is,  to check/run migration when the app is running, instead of use lucky binary.

Then, i add following code into the last line of a new 1.1.0 project's config/database.cr

```
if !LuckyEnv.task?
  Avram::Migrator::Runner.new.run_pending_migrations
end
```

I found those code never work, because `Avram::Migrator::Runner.migrations` always return a empty list, l always get empty migrations.

```
sorted_migrations # => self.class.migrations # => []
[]
self.class.migrations # => []
Did not migrate anything because there are no pending migrations.
```

Maybe because I haven't used the Lucky/avram for a while now, but, i remember when i start o use avram with kemal,  i can achieve this.

I guess, maybe i missing some config which search migrations file in `db/migrations` folder? there is `00000000000001_create_users.cr` there, i still not create user table yet.

Thanks!

**Comment by zw963** [comment, marked as answer]
Okay, add a `require "../db/migrations/**"` into `config/database.cr` do the trick.

So, is there possible to create a precompile binary, like `bin/lucky.db.migrate`? 

Thanks.

**Reply by robacarp** [reply]
A lucky app provides tasks.cr, which is what eventually gets executed when you call `lucky db.migrate` from your terminal. If you add it to your shard.yml as a target:

```yaml
targets:
  tasks:
    main: tasks.cr
```

Then when you deploy with a `shards build`, you'll get an executable at bin/tasks which you can use to run migrations with `bin/tasks db.migrate`

**Comment by zw963** [comment]
> Okay, add a require "../db/migrations/**" into config/database.cr do the trick.

Hi, i know the reason i have to add `require "../db/migration/**"` into `config/database.cr` for solve my issue. 

because the require order in the `src/app.cr` cause this issue.

```
require "./shards"

# ...
require "../config/**"

# ...
require "../db/migrations/**"
require "./app_server"
```

Is there special reason we can't write the  "require ../db/migrations/* *" earlier? just before the `require "../config"`?


Or,  just like how I have to manually modify every time, add it into require "../config/database.cr"?

```diff
+ require "../db/migrations/**"

# ...

+ if !LuckyEnv.task?
+   Avram::Migrator::Runner.new.run_pending_migrations
+ end
```

**Reply by jwoertink** [reply]
I think as long as your migrations are required after your models, it should be fine. 

```crystal
# the migration would need to know that Post and User exists
create table_for(Post) do
  # ...
  add_belongs_to user : User
end
```

But since those are used in a Macro, and not actually `resolve`d, then maybe you could even move the require to after `require "avram"`?

**Reply by zw963** [reply]
> But since those are used in a Macro, and not actually resolved, then maybe you could even move the require to after require "avram"?

I think so too.

The issue is, where to add `Avram::Migrator::Runner.new.run_pending_migrations`, okay, please let me explain why i need check pending migrations in my deployment process. (anyway, I think the process introduced in official doc is a bit complicated)

1.  yard proc for package assets into public folder
2.  use [baked_file_system_mounter](https://github.com/crystal-china/baked_file_system_mounter) to add assets files in public into binary when build, and auto mount it on production.
3.  build static locally use zig cc. there is a [toturial](https://github.com/zw963/crystal-book/blob/patch-1/docs/guides/use_zig_cc_as_an_alternative_linker.md) for how to do this, then i have a big static binary, which is ready for copy to production.
4. copy file into production, and run it, done.
5.  it will mount assets file into specified folder (public),  and check the pending migration, run migration if necessary.  no need crystal or nodejs installed on server, deploy is easy.

The only issue is, where i add `Avram::Migrator::Runner.new.run_pending_migrations` into? the prefer place is `config/database.cr`, but add to that end of that file not work, so, i have to add `require "../db/migrations/**"`  at the beginning of `config/database.cr` too solve this issue.

In fact, i tried, remove `require "../db/migrations/**"` ` from `src/app.cr`, and add those back into `config/database.cr`, it work, and this refactor seem like make sense.

Thanks

**Reply by jwoertink** [reply]
All of my apps use `Avram::Migrator::Runner.new.run_pending_migrations` on deploy too. I put my require migrations in `app.cr` and the method call goes in `start_server.cr`

```
require "./app"

Habitat.raise_if_missing_settings!

if LuckyEnv.development?
  Avram::Migrator::Runner.new.ensure_migrated!
  Avram::SchemaEnforcer.ensure_correct_column_mappings!
else
  Avram::Migrator::Runner.new.run_pending_migrations
end

app_server = AppServer.new

Signal::INT.trap do
  app_server.close
end

app_server.listen
```

**Comment by zw963** [comment]
> I put my require migrations in `app.cr`

Yes, i thought this is the key point. migration is related to database, add it into config/database.cr is better?

Or, if anyone use avaram without lucky, where the `run_pending_migrations` should add to?  i guess still should be config/database.cr

Thanks
